pkg/reconciler/knativeserving/ingress/kourier_test.go

```diff
@@ -19,85 +19,116 @@ package ingress
+		expNamespace:   "kourier-system", // kourier default namespace
+		expServiceType: "LoadBalancer",   // kourier GW default service type
 	}}
 
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			u := makeUnstructuredDeployment(t, tt.deploymentName, tt.ns, tt.labels)
-			replaceKourierGWNamespace(tt.expectedNS)(u)
-			deployment := &appsv1.Deployment{}
-			err := scheme.Scheme.Convert(u, deployment, nil)
-			util.AssertEqual(t, err, nil)
-			util.AssertDeepEqual(t, deployment.Spec.Template.Spec.Containers, tt.expected)
+			client := fake.New()
+			manifest, err := mf.NewManifest("testdata/kodata/ingress/0.20/kourier.yaml", mf.UseClient(client))
+			if err != nil {
+				t.Fatalf("Failed to read manifest: %v", err)
+			}
+
+			if tt.dropLabel {
+				manifest, err = manifest.Transform(removeLabels())
+				if err != nil {
+					t.Fatalf("Failed to transform manifest: %v", err)
+				}
+			}
+
+			manifest, err = manifest.Transform(replaceGWNamespace(tt.instance))
+			if err != nil {
+				t.Fatalf("Failed to transform manifest: %v", err)
+			}
+
+			manifest, err = manifest.Transform(configureGWServiceType(tt.instance))
+			if err != nil {
+				t.Fatalf("Failed to transform manifest: %v", err)
+			}
+
+			for _, u := range manifest.Resources() {
+				if u.GetKind() == "Deployment" && u.GetName() == kourierControllerDeploymentName {
+					deployment := &appsv1.Deployment{}
+					err := scheme.Scheme.Convert(&u, deployment, nil)
+					util.AssertEqual(t, err, nil)
+					envs := deployment.Spec.Template.Spec.Containers[0].Env
+					env := ""
+					for i := range envs {
+						if envs[i].Name == kourierGatewayNSEnvVarKey {
+							env = envs[i].Value
+						}
+					}
+					util.AssertDeepEqual(t, env, tt.expNamespace)
+				}
+				if u.GetKind() == "Service" && u.GetName() == kourierGatewayServiceName {
+					svc := &v1.Service{}
+					err := scheme.Scheme.Convert(&u, svc, nil)
+					util.AssertEqual(t, err, nil)
+					svcType := svc.Spec.Type
+					util.AssertDeepEqual(t, string(svcType), tt.expServiceType)
+				}
+			}
 		})
 	}
 }
 
-func makeUnstructuredDeployment(t *testing.T, name, ns string, labels map[string]string) *unstructured.Unstructured {
-	d := &appsv1.Deployment{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:   name,
-			Labels: labels,
-		},
-		Spec: appsv1.DeploymentSpec{
-			Template: v1.PodTemplateSpec{
-				Spec: v1.PodSpec{
-					Containers: []v1.Container{{
-						Name: name,
-						Env:  []v1.EnvVar{{Name: kourierGatewayNSEnvVarKey, Value: ns}},
-					}},
-				},
-			},
-		},
-	}
-	result := &unstructured.Unstructured{}
-	err := scheme.Scheme.Convert(d, result, nil)
-	if err != nil {
-		t.Fatalf("Could not create unstructured Deployment: %v, err: %v", d, err)
+// removeProviderLabels removes labels. This util is used for tests without provider label.
+func removeLabels() mf.Transformer {
+	return func(u *unstructured.Unstructured) error {
+		u.SetLabels(map[string]string{})
+		return nil
```

Comment:
I don’t understand why this always return nil. Why not make it a void function instead?

Reply:
The `removeLabels()` is used `manifest.Transform()` and the type is defined in:

https://github.com/knative/operator/blob/846d8762f2d1df4a0d080e30ff89c54cf845be02/vendor/github.com/manifestival/manifestival/transform.go#L13

And it needs to return `error `.